# **Kaggle Challenge 2: Đề tài dự báo giá nhà.**
# **PHẦN 2: PHÂN TÍCH KHÁM PHÁ DỮ LIỆU**

## 1. Mô tả vấn đề
+ **Mô tả**:
   - Dự báo giá nhà dựa trên 79 tính chất của dataset nhà ở tại Ames, Iowa.
   - Dữ liệu đầu vào gồm 2 file:
      * train_clean.pkl: Dữ liệu từ file train đã được xử lý ở phần 1.
      * test_clean.pkl: Dữ liệu từ file test đã được xử lý ở phần 1.

+ **Mục tiêu**:
   - Phân tích và khám phá, tìm đặc trưng của các cột trong dữ liệu.



## 2. Chuẩn bị                

### 2.1. Import các thư viện cần thiết

In [743]:
# Load libraries
from IPython import display
import numpy as np
import pickle

import matplotlib.pyplot as plt

import pandas as pd
import seaborn as sns


### 2.2. Đọc dataset train và test từ dataset đã xử lý trong phần dọn dẹp dữ liệu
Đọc các file .pkl trong thư mục /clean/data

In [744]:
train_data = pd.DataFrame(pd.read_pickle("../clean/data/train_clean.pkl"))
test_data = pd.DataFrame(pd.read_pickle("../clean/data/test_clean.pkl"))

### 2.3. Kiểm tra dữ liệu dataset train và test

In [745]:
train_data.head()

,Id,MSSubClass,MSZoning,LotFrontage,LotArea,Street,Alley,LotShape,LandContour,Utilities,...,PoolArea,PoolQC,Fence,MiscFeature,MiscVal,MoSold,YrSold,SaleType,SaleCondition,SalePrice
0,1,60,RL,65,8450,Pave,None,Reg,Lvl,AllPub,...,0,None,None,None,0,2,2008,WD,Normal,208500
1,2,20,RL,80,9600,Pave,None,Reg,Lvl,AllPub,...,0,None,None,None,0,5,2007,WD,Normal,181500
2,3,60,RL,68,11250,Pave,None,IR1,Lvl,AllPub,...,0,None,None,None,0,9,2008,WD,Normal,223500
3,4,70,RL,60,9550,Pave,None,IR1,Lvl,AllPub,...,0,None,None,None,0,2,2006,WD,Abnorml,140000
4,5,60,RL,84,14260,Pave,None,IR1,Lvl,AllPub,...,0,None,None,None,0,12,2008,WD,Normal,250000


In [746]:
test_data.head()

,Id,MSSubClass,MSZoning,LotFrontage,LotArea,Street,Alley,LotShape,LandContour,Utilities,...,ScreenPorch,PoolArea,PoolQC,Fence,MiscFeature,MiscVal,MoSold,YrSold,SaleType,SaleCondition
0,1461,20,RH,80,11622,Pave,None,Reg,Lvl,AllPub,...,120,0,None,MnPrv,None,0,6,2010,WD,Normal
1,1462,20,RL,81,14267,Pave,None,IR1,Lvl,AllPub,...,0,0,None,None,Gar2,12500,6,2010,WD,Normal
2,1463,60,RL,74,13830,Pave,None,IR1,Lvl,AllPub,...,0,0,None,MnPrv,None,0,3,2010,WD,Normal
3,1464,60,RL,78,9978,Pave,None,IR1,Lvl,AllPub,...,0,0,None,None,None,0,6,2010,WD,Normal
4,1465,120,RL,43,5005,Pave,None,IR1,HLS,AllPub,...,144,0,None,None,None,0,1,2010,WD,Normal


In [747]:
print("Số cột null trong train: ", train_data.isnull().sum().gt(0).sum())
print("Số cột null trong test: ", test_data.isnull().sum().gt(0).sum())

Số cột null trong train:  0
Số cột null trong test:  0


In [748]:
train_data.dtypes.value_counts()

object    43
int64     38
Name: count, dtype: int64

In [749]:
test_data.dtypes.value_counts()

object    43
int64     37
Name: count, dtype: int64

## 3. Phân tích khai phá dữ liệu

### 3.1. Phân tích đơn biến

#### **1) Các cột số**

##### a. Phân phối các cột số theo biểu đồ cột (Histogram Plot)

In [750]:
# # Tạo lưới biểu đồ
# fig, axes = plt.subplots(10, 4, figsize=(20, 30))
# axes = axes.flatten()

# # Lặp qua các cột trong dataframe
# for i, col in enumerate(train_data.select_dtypes(include=['int64']).columns):
#     sns.histplot(train_data[col], bins=50, kde=True, color='skyblue', ax=axes[i])
#     axes[i].set_title(col, fontsize=10)
#     axes[i].set_xlabel('')
#     axes[i].set_ylabel('')

# # Xóa các subplot không dùng đến (dư thừa)
# for j in range(len(train_data.select_dtypes(include=['int64']).columns), len(axes)):
#     fig.delaxes(axes[j])

# plt.tight_layout()
# plt.savefig("fig_hist_1a.png")
# plt.show()

**Nhận xét**
- Cột Id chứa các giá trị duy nhất.
- Các cột MSSubClass, BsmtFullBath, BsmtHalfBath, FullBath, HalfBath, BedroomAbvGr, KitchenAbvGr, 
TotRmsAbvGrd, Fireplaces, GarageCars, MoSold và YrSold có phân bố rời rạc.
- Các cột LotFrontage, LotArea, MasVnrArea, BsmtFinSF1, BsmtFinSF2, BsmtUnfSF, TotalBsmtSF, 1stFlrSF,2ndFlrSF, LowQualFinSF, GrLivArea, GarageArea, WoodDeckSF, OpenPorchSF, EnclosedPorch, 3SsnPorch, ScreenPorch, PoolArea và MiscVal có phân bố liên tục.
- Các cột YearBuilt, YearRemodAdd, GarageYrBlt, MoSold và YrSold là các cột thứ tự thời gian.
- Các cột OverallQual, OverallCond  là các cột có tính chất phân loại (đánh giá).

##### b. Phân phối các cột có phân bố liên tục theo biểu đồ hộp (Box plot)

In [751]:
# continous_cols = [
#     "LotFrontage", "LotArea", "MasVnrArea", "BsmtFinSF1", "BsmtFinSF2",
#     "BsmtUnfSF", "TotalBsmtSF", "1stFlrSF", "2ndFlrSF", "LowQualFinSF",
#     "GrLivArea", "GarageArea", "WoodDeckSF", "OpenPorchSF", "EnclosedPorch",
#     "3SsnPorch", "ScreenPorch", "PoolArea", "MiscVal"
# ]
# # Tạo lưới biểu đồ
# fig, axes = plt.subplots(5, 4, figsize=(20, 20))
# axes = axes.flatten()

# # Lặp qua các cột trong dataframe
# for i, col in enumerate(train_data[continous_cols]):
#     sns.boxplot(train_data[col], color="skyblue", fliersize=3, linewidth=1, ax=axes[i])
#     axes[i].set_title(col, fontsize=10)
#     axes[i].set_xlabel('')
#     axes[i].set_ylabel('')

# # Xóa các subplot không dùng đến (dư thừa)
# for j in range(len(continous_cols), len(axes)):
#     fig.delaxes(axes[j])

# plt.tight_layout()
# plt.savefig("fig_box_1a.png")
# plt.show()

**Nhận xét**
+ Các cột MasVnrArea, BsmtFinSF1, BsmtFinSF2, 2ndFlrSF, LowQualFinSF, WoodDeckSF, OpenPorchSF, EnclosedPorch, 3SsnPorch, 
ScreenPorch, PoolArea, MiscVal có phân bố lệch phải mạnh với nhiều giá trị ngoại lai (có phân phối hình chữ L). Giá trị
tập trung hầu hết ở 0 (0 ở đây có khả năng là không có).
+ Các cột LotArea, LotFrontage, BsmtUnfSF, TotalBsmtSF, 1stFlrSF, GrLivArea, GarageArea có phân bố ổn định hơn nhưng vẫn lệch phải với nhiều giá trị ngoại lai. 

##### c. Xem phân phối giá trị các cột phân phối rời rạc

In [752]:
discrete_cols = [
    "MSSubClass", 
    "BsmtFullBath", 
    "BsmtHalfBath", 
    "FullBath", 
    "HalfBath", 
    "BedroomAbvGr", 
    "KitchenAbvGr", 
    "TotRmsAbvGrd", 
    "Fireplaces", 
    "GarageCars"
]

for col in discrete_cols:
      print("-------------------------------")
      print(f"Phân phối cột {col} (train): ", train_data[col].sort_values().unique(), ", có", 
            train_data[col].unique().size, " giá trị ")
      print(f"Phân phối cột {col} (test): ", test_data[col].sort_values().unique(), ", có", 
            test_data[col].unique().size, " giá trị ")



-------------------------------
Phân phối cột MSSubClass (train):  [ 20  30  40  45  50  60  70  75  80  85  90 120 160 180 190] , có 15  giá trị 
Phân phối cột MSSubClass (test):  [ 20  30  40  45  50  60  70  75  80  85  90 120 150 160 180 190] , có 16  giá trị 
-------------------------------
Phân phối cột BsmtFullBath (train):  [0 1 2 3] , có 4  giá trị 
Phân phối cột BsmtFullBath (test):  [0 1 2 3] , có 4  giá trị 
-------------------------------
Phân phối cột BsmtHalfBath (train):  [0 1 2] , có 3  giá trị 
Phân phối cột BsmtHalfBath (test):  [0 1 2] , có 3  giá trị 
-------------------------------
Phân phối cột FullBath (train):  [0 1 2 3] , có 4  giá trị 
Phân phối cột FullBath (test):  [0 1 2 3 4] , có 5  giá trị 
-------------------------------
Phân phối cột HalfBath (train):  [0 1 2] , có 3  giá trị 
Phân phối cột HalfBath (test):  [0 1 2] , có 3  giá trị 
-------------------------------
Phân phối cột BedroomAbvGr (train):  [0 1 2 3 4 5 6 8] , có 8  giá trị 
Phân phối cột Bed

**Nhận xét**: 
- Cột MSSubClass có 16 loại giá trị trên test, nhưng trong train chỉ có 15 giá trị(thiếu giá trị 150).
- Cột FullBath có 5 loại giá trị trên test, nhưng trong train chỉ có 4 giá trị(thiếu giá trị 4).
- Cột BedroomAbvGr có 7 loại giá trị trên test, nhưng trong train có 8 giá trị (dư giá trị 8).
- Cột KitchenAbvGr có 3 loại giá trị trên test, nhưng trong train có 4 giá trị (dư giá trị 3).
- Cột TotRmsAbvGrd có 12 loại giá trị trên test, trong train có 12 giá trị nhưng phân bố khác với test.
- Cột Fireplaces có 5 loại giá trị trên test, nhưng trong train chỉ có 4 giá trị (thiếu giá trị 4).
- Cột GarageCars có 6 loại giá trị trên test, nhưng trong train có 6 loại giá trị (thiếu giá trị 5).
- Cột BsmtFullBath, BsmtHalfBath, HalfBath có phân phối giống nhau trên cả train và test.

**Giá trị trong hầu hết các cột trừ MSSubClass đều biểu hiện quan hệ thứ tự hay khoảng cách nên không cần biến đổi. Do đó, cột MSSubClass cần được xử lý riêng (do sự phân bố của nó được đề cập trong file mô tả).**

##### d. Xem phân phối của các cột phân loại (đánh giá)

In [753]:
for col in ["OverallQual", "OverallCond"]:
    print("-------------------------------")
    print(f"Phân phối cột {col} (train): ", train_data[col].sort_values().unique(), ", có", 
        train_data[col].unique().size, " giá trị ")
    print(f"Phân phối cột {col} (test): ", test_data[col].sort_values().unique(), ", có", 
        test_data[col].unique().size, " giá trị ")



-------------------------------
Phân phối cột OverallQual (train):  [ 1  2  3  4  5  6  7  8  9 10] , có 10  giá trị 
Phân phối cột OverallQual (test):  [ 1  2  3  4  5  6  7  8  9 10] , có 10  giá trị 
-------------------------------
Phân phối cột OverallCond (train):  [1 2 3 4 5 6 7 8 9] , có 9  giá trị 
Phân phối cột OverallCond (test):  [1 2 3 4 5 6 7 8 9] , có 9  giá trị 


**Nhận xét**: Cả 2 cột OverallQual và OverallCond đều phân phối giá trị giống nhau trên train và test

##### e. Xem phân phối các cột thứ tự thời gian

In [754]:
time_cols = ["YearBuilt", "YearRemodAdd", "GarageYrBlt", "MoSold", "YrSold"]

for col in time_cols:
    print("-------------------------------")
    
    train_vals = set(train_data[col].unique())
    test_vals = set(test_data[col].unique())
    
    print(f"Phân phối cột {col} (train): {sorted(train_vals)}, có {len(train_vals)} giá trị")
    print(f"Phân phối cột {col} (test): {sorted(test_vals)}, có {len(test_vals)} giá trị")
    
    # Giá trị có trong test nhưng không có trong train
    print(f"Giá trị chỉ có trong test: {sorted(test_vals - train_vals)}")
    
    # Giá trị có trong train nhưng không có trong test
    print(f"Giá trị chỉ có trong train: {sorted(train_vals - test_vals)}")

-------------------------------
Phân phối cột YearBuilt (train): [np.int64(1872), np.int64(1875), np.int64(1880), np.int64(1882), np.int64(1885), np.int64(1890), np.int64(1892), np.int64(1893), np.int64(1898), np.int64(1900), np.int64(1904), np.int64(1905), np.int64(1906), np.int64(1908), np.int64(1910), np.int64(1911), np.int64(1912), np.int64(1913), np.int64(1914), np.int64(1915), np.int64(1916), np.int64(1917), np.int64(1918), np.int64(1919), np.int64(1920), np.int64(1921), np.int64(1922), np.int64(1923), np.int64(1924), np.int64(1925), np.int64(1926), np.int64(1927), np.int64(1928), np.int64(1929), np.int64(1930), np.int64(1931), np.int64(1932), np.int64(1934), np.int64(1935), np.int64(1936), np.int64(1937), np.int64(1938), np.int64(1939), np.int64(1940), np.int64(1941), np.int64(1942), np.int64(1945), np.int64(1946), np.int64(1947), np.int64(1948), np.int64(1949), np.int64(1950), np.int64(1951), np.int64(1952), np.int64(1953), np.int64(1954), np.int64(1955), np.int64(1956), np.int

**Nhận xét**: 
- Các cột YearRemodAdd, MoSold, YrSold có phân phối giá trị giống nhau.
- Các cột YearBuilt, GarageYrBuilt biểu hiện quan hệ thứ tự, do đó không cần mã hóa One Hot.

#### **2) Các cột phi số**